internal/observability/events/event_audit.go

```diff
@@ -0,0 +1,142 @@
+	Version        string       `json:"version"`                // std audit/boundary field
+	Type           string       `json:"type"`                   // std audit field
+	Timestamp      time.Time    `json:"timestamp"`              // std audit field
+	RequestInfo    *RequestInfo `json:"request_info,omitempty"` // boundary field
+	Auth           *Auth        `json:"auth,omitempty"`         // std audit field
+	Request        *Request     `json:"request,omitempty"`      // std audit field
+	Response       *Response    `json:"response,omitempty"`     // std audit field
+	SerializedHMAC string       `json:"serialized_hmac"`        // boundary field
+	Flush          bool         `json:"-"`
+}
+
+func newAudit(fromOperation Op, opt ...Option) (*Audit, error) {
+	const op = "event.newAudit"
+	if fromOperation == "" {
+		return nil, errors.New(errors.InvalidParameter, op, "missing from operation")
+	}
+	opts := getOpts(opt...)
+	if opts.withId == "" {
+		var err error
+		opts.withId, err = newId(string(AuditType))
+		if err != nil {
+			return nil, errors.Wrap(err, op)
+		}
+	}
+	var dtm time.Time
+	switch opts.withNow.IsZero() {
+	case false:
+		dtm = opts.withNow
+	default:
+		dtm = time.Now()
+	}
+
+	a := &Audit{
+		Id:          opts.withId,
+		Version:     AuditVersion,
+		Type:        string(ApiRequest),
+		Timestamp:   dtm,
+		RequestInfo: opts.withRequestInfo,
+		Auth:        opts.withAuth,
+		Request:     opts.withRequest,
+		Response:    opts.withResponse,
+		Flush:       opts.withFlush,
+	}
+	if err := a.validate(); err != nil {
+		return nil, errors.Wrap(err, op)
+	}
+	return a, nil
+}
+
+// EventType is required for all event types by the eventlogger broker
+func (a *Audit) EventType() string { return string(AuditType) }
+
+func (a *Audit) validate() error {
+	const op = "event.(audit).validate"
+	if a.Id == "" {
+		return errors.New(errors.InvalidParameter, op, "missing id")
+	}
+	return nil
+}
+
+// GetID is part of the eventlogger.Gateable interface and returns the audit
+// event's id.
+func (a *Audit) GetID() string {
+	return a.Id
+}
+
+// FlushEvent is part of the eventlogger.Gateable interface and returns the
+// value of the audit event's flush field
+func (a *Audit) FlushEvent() bool {
+	return a.Flush
+}
+
+// ComposedFrom is part of the eventlogger.Gatable interface.  It's important to
+// remember that the receiver will always be nil when this is called by the eventlogger.GatedFilter
+func (a *Audit) ComposeFrom(events []*eventlogger.Event) (eventlogger.EventType, interface{}, error) {
+	const op = "event.(audit).ComposedFrom"
+	if len(events) == 0 {
+		return "", nil, errors.New(errors.InvalidParameter, op, "missing events")
+	}
+	var validId string
+	var initValidId sync.Once
+	payload := Audit{}
+	for i, v := range events {
+		gated, ok := v.Payload.(*Audit)
+		if !ok {
+			return "", nil, errors.New(errors.InvalidParameter, op, fmt.Sprintf("event %d is not an audit payload", i))
+		}
+		initValidId.Do(func() {
```

Comment:
Curious what the intent here is for using `sync.Once` to initialize this - only thing I can think of is to protect against concurrency, but I don't see anything here that would need that? Would a simple empty check/non-empty assertion on `gated.Id` be better?

```go
if gated.Id == "" {
  return "", nil, errors.New(errors.InvalidParameter, op, fmt.Sprintf("event %d: Id is required", i))
}

if validId == {
  validId = gated.Id
}
```

Reply:
ty!